Improve label parsing

jrajahalme opened the pull request:
Fix parsing of labels like 'reservedz=foo".

Use strings.IndexByte() instead of strings.SplitN() when the delimiter is a single character and N==2. This avoids allocating string slices for the SplitN() return values, which has shown up in pprof output.

These changes make `labels.ParseLabel()` about 4x faster than before.

```release-note
Improve label parsing to be 4x faster.
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5928)
<!-- Reviewable:end -->

aanm commented on pkg/labels/filter.go:
```diff
@@ -85,10 +85,10 @@ func (p LabelPrefix) matches(l *Label) (bool, int) {
-		labelPrefix.Source = t[0]
-		labelPrefix.Prefix = t[1]
+	i := strings.IndexByte(label, ':')
+	if i >= 0 {
```
shouldn't this be `i > 0`?

jrajahalme replied:
`IndexByte()` returns non-negative index when the byte is found. `SplitN()` returns 2 if the delimiter is found. This test was `> 1` before, so the corresponding test is `>= 0`. Are you saying that an empty source like ":prefix" should be parsed as `{Source="", Prefix=":prefix"}` instead of `{Source="", Prefix="prefix"}`? If so this would be a bug fix that I'd separate to a commit of it's own.

aanm replied:
> `IndexByte()` returns non-negative index when the byte is found. `SplitN()` returns 2 if the delimiter is found. 

I forgot that.

>This test was `> 1` before, so the corresponding test is `>= 0`. Are you saying that an empty source like ":prefix" should be parsed as `{Source="", Prefix=":prefix"}` instead of `{Source="", Prefix="prefix"}`? 

No

aanm commented on pkg/labels/filter.go:
```diff
@@ -85,10 +85,10 @@ func (p LabelPrefix) matches(l *Label) (bool, int) {
+	i := strings.IndexByte(label, ':')
+	if i >= 0 {
+		labelPrefix.Source = label[:i]
+		labelPrefix.Prefix = label[i+1:]
```
This will panic if I execute `parseLabelPrefix(":")`

jrajahalme replied:
In that case `Source = ":"[:0]`, which is `""`, and `Prefix = ":"[1:]`, which is `""`, no panic. I added a test case for `":"` in to TestParseLabel(), which has the equivalent computation to be absolutely sure.

aanm commented on pkg/labels/labels.go:
```diff
@@ -625,13 +621,15 @@ func ParseLabel(str string) *Label {
+		lbl.Key = next
+	} else {
+		if i == 0 && src == LabelSourceReserved {
+			lbl.Key = next[i+1:]
```
This line is not covered by a unit test

jrajahalme replied:
Added a unit test line that should cover this, but not sure if this makes sense:
```
		{"reserved:=key", NewLabel("key", "", LabelSourceReserved)},
```

aanm replied:
should be `{"reserved:key", NewLabel("key", "", LabelSourceReserved)},`?

jrajahalme replied:
Test like that already exists, but it does not cover the lines you commented about. The code branch that was not tested before explicitly tests for an empty key AND "reserved" label source.  Note that this PR does not change the behavior of these two lines, only their implementation.

aanm commented on pkg/labels/filter.go:
```diff
@@ -85,10 +85,10 @@ func (p LabelPrefix) matches(l *Label) (bool, int) {
-		labelPrefix.Prefix = t[1]
+	i := strings.IndexByte(label, ':')
+	if i >= 0 {
+		labelPrefix.Source = label[:i]
```
These 2 lines are not covered by a unit test

jrajahalme replied:
Changed an existing test to cover these lines.

jrajahalme committed:
labels: Fix parsing of reserved labels.

Fix parsing of reserved labels so that, e.g., "reservedz=foo" is not
considered having a reserved source.

Also change unit tests to cover more of the sources.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
benchmark

jrajahalme committed:
labels: Open code strings.SplitN(,,2)

We only split into two using a single character delimiter, so it is
possible to find the delimiter with strings.IndexByte() instead of
using the more generic strings.SplitN(). SplitN() always allocates a
new slice of strings for the return value, so using IndexByte() is
also faster and more GC friendly. Based on the benchmark test
ParseLabels() is now about 4x faster than before.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aanm commented on pkg/labels/filter_test.go:
```diff
@@ -33,7 +33,7 @@ func (s *LabelsPrefCfgSuite) TestFilterLabels(c *C) {
 
 	dlpcfg := defaultLabelPrefixCfg()
-	d, err := parseLabelPrefix("!ignor[eE]")
+	d, err := parseLabelPrefix(":!ignor[eE]")
```
Is this change necessary?

jrajahalme replied:
It is, to cover the if branch you said is not covered by the unit tests.

aanm replied:
I thought you were accidentally changing behavior.

jrajahalme replied:
My bad, I should have called out this change in the commit message.
